spi_nand_flash/src/nand.c

```diff
@@ -3,172 +3,24 @@
-        dev->dhara_nand.log2_page_size = 12; // 4k pages
-        dev->dhara_nand.num_blocks = 4096;
-        dev->read_page_delay_us = 130; // somewhat slower reads
-        break;
-    default:
-        return ESP_ERR_INVALID_RESPONSE;
-    }
-    return ESP_OK;
-}
-
-static esp_err_t spi_nand_gigadevice_init(spi_nand_flash_device_t *dev)
-{
-    uint8_t device_id;
-    spi_nand_transaction_t t = {
-        .command = CMD_READ_ID,
-        .dummy_bits = 16,
-        .miso_len = 1,
-        .miso_data = &device_id,
-        .flags = SPI_TRANS_USE_RXDATA,
-    };
-    spi_nand_execute_transaction(dev->config.device_handle, &t);
-    dev->read_page_delay_us = 25;
-    dev->erase_block_delay_us = 3200;
-    dev->program_page_delay_us = 380;
-    ESP_LOGD(TAG, "%s: device_id: %x\n", __func__, device_id);
-    switch (device_id) {
-    case GIGADEVICE_DI_51:
-    case GIGADEVICE_DI_41:
-    case GIGADEVICE_DI_31:
-    case GIGADEVICE_DI_21:
-        dev->dhara_nand.num_blocks = 1024;
-        break;
-    case GIGADEVICE_DI_52:
-    case GIGADEVICE_DI_42:
-    case GIGADEVICE_DI_32:
-    case GIGADEVICE_DI_22:
-        dev->dhara_nand.num_blocks = 2048;
-        break;
-    case GIGADEVICE_DI_55:
-    case GIGADEVICE_DI_45:
-    case GIGADEVICE_DI_35:
-    case GIGADEVICE_DI_25:
-        dev->dhara_nand.num_blocks = 4096;
-        break;
-    default:
-        return ESP_ERR_INVALID_RESPONSE;
-    }
-    return ESP_OK;
-}
-
-static esp_err_t spi_nand_micron_init(spi_nand_flash_device_t *dev)
-{
-    uint8_t device_id;
-    spi_nand_transaction_t t = {
-        .command = CMD_READ_ID,
-        .dummy_bits = 16,
-        .miso_len = 1,
-        .miso_data = &device_id,
-        .flags = SPI_TRANS_USE_RXDATA,
-    };
-    spi_nand_execute_transaction(dev->config.device_handle, &t);
-    dev->ecc_data.ecc_status_reg_len_in_bits = 3;
-    dev->erase_block_delay_us = 2000;
-    ESP_LOGD(TAG, "%s: device_id: %x\n", __func__, device_id);
-    switch (device_id) {
-    case MICRON_DI_34:
-        dev->read_page_delay_us = 115;
-        dev->program_page_delay_us = 240;
-        dev->dhara_nand.num_blocks = 2048;
-        dev->dhara_nand.log2_ppb = 6;        // 64 pages per block
-        dev->dhara_nand.log2_page_size = 12; // 4096 bytes per page
-        break;
-    case MICRON_DI_14:
-    case MICRON_DI_15:
-        dev->read_page_delay_us = 46;
-        dev->program_page_delay_us = 220;
-        dev->dhara_nand.num_blocks = 1024;
-        dev->dhara_nand.log2_ppb = 6;          // 64 pages per block
-        dev->dhara_nand.log2_page_size = 11;   // 2048 bytes per page
-        break;
-    default:
-        return ESP_ERR_INVALID_RESPONSE;
-    }
-    return ESP_OK;
-}
+extern esp_err_t spi_nand_winbond_init(spi_nand_flash_device_t *dev);
+extern esp_err_t spi_nand_alliance_init(spi_nand_flash_device_t *dev);
+extern esp_err_t spi_nand_gigadevice_init(spi_nand_flash_device_t *dev);
+extern esp_err_t spi_nand_micron_init(spi_nand_flash_device_t *dev);
```

Comment:
Can we create a private header file and declare these functions there, instead of using an extern declaration?

The advantage of the common header file is that it will prevent accidentally having incompatible declarations.

Reply:
Yes, that makes sense.